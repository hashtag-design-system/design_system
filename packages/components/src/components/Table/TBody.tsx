import React from "react";
import { useClassnames } from "../../utils";

export type FProps = React.ComponentProps<"tbody">;

const TBody: React.FC<FProps> = ({ children, ...props }) => {
  const [classNames, rest] = useClassnames("table__tbody", props);

  return (
    <tbody className={classNames} data-testid="table-tbody" {...rest}>
      {children}
    </tbody>
  );
};

TBody.displayName = "TableTBody";

export default TBody;
