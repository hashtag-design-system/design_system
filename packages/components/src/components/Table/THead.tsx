import React from "react";
import { useClassnames } from "../../utils/hooks";

export type FProps = React.ComponentProps<"thead">

const THead: React.FC<FProps> = ({ children, ...props }) => {
  const [classNames, rest] = useClassnames("table__thead", props);

  return (
    <thead className={classNames} data-testid="table-thead" {...rest}>
      {children}
    </thead>
  );
};

THead.displayName = "TableTHead";

export default THead;
