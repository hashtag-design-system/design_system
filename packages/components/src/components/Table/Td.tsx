import React from "react";
import { useClassnames } from "../../utils";

export type FProps = React.ComponentProps<"td">;

const Td: React.FC<FProps> = ({ children, ...props }) => {
  const [classNames, rest] = useClassnames("table__tbody__td", props);

  return (
    <td className={classNames} data-testid="table-td" {...rest}>
      {children}
    </td>
  );
};

Td.displayName = "TableTd";

export default Td;
