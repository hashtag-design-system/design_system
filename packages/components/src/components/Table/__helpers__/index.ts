export { ChevronDown } from "./ChevronDown";
export { ChevronUp } from "./ChevronUp";
export { ChevronUpAndDown } from "./ChevronUpAndDown";
