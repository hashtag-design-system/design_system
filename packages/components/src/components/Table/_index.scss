.table {
  --border-radius: 8px;
  $border: 1px solid $grey-4;
  margin: 0;
  padding: 0;
  list-style: none;
  width: min(100%, 75vw);
  text-align: left;
  border-collapse: collapse;
  border-spacing: 0;

  &-border {
    border-bottom: $border;
  }

  &__thead {
    @extend .body-12;
    @extend .table-border;
    border-width: 2px;
    width: 100%;
    height: auto;
    background-color: $grey-3;

    &__th {
      @extend .bold;
      @extend .upper;
      height: 100%;
      padding: 1.25em 1.25em;
      color: $grey-9;

      &:first-of-type {
        border-top-left-radius: var(--border-radius);
      }

      &:last-of-type {
        border-top-right-radius: var(--border-radius);
      }

      &:not([aria-sort="none"]) {
        cursor: pointer;

        div {
          justify-content: space-between;
        }
      }

      div {
        @extend .flex-row-flex-start-center;
        gap: 8px;
        height: 50px;
        height: 100%;
      }
    }

    .table__tr {
      height: em(40);
    }
  }

  &__tbody {
    @extend .body-14;
    // Due to <Checkbox />'s height
    line-height: em(24) !important;

    &__td {
      padding: 0.75em 1em;
      text-align: left;
    }

    .table__tr {
      @include hover-classNames() {
        background-color: lighten($secondary, 47.5);
      }
    }
  }

  &__tr {
    @extend .table-border;
    height: em(36);
    will-change: background-color;
    border-spacing: 0;

    // Capture both in <thead /> and <tbody />
    input[type="checkbox"],
    input[type="radio"] {
      box-shadow: none;
    }

    &.border-right {
      td,
      th {
        &:first-child {
          border-right: $border;
        }
      }
    }
  }
}
