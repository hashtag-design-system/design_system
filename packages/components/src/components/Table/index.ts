import Table from "./Table";

export type { FProps as TableFProps, Props as TableProps } from "./Table";
export type { FProps as TableTBodyFProps } from "./TBody";
export type { FProps as TableTdFProps } from "./Td";
export type { FProps as TableThFProps, Props as TableThProps } from "./Th";
export type { FProps as TableTHeadFProps } from "./THead";
export { TrStates as TableTrStates } from "./Tr";
export type { FProps as TableTrFProps, Props as TableTrProps, TrStateType as TableTrStateType } from "./Tr";

export default Table;
