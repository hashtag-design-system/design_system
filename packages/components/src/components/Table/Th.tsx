import React, { useState } from "react";
import { SortDirection } from "../../typings";
import { useClassnames } from "../../utils";
import { ChevronDown, ChevronUp, ChevronUpAndDown } from "./__helpers__";

export type Props = {
  sort?: boolean;
  onClick?: (e: React.MouseEvent<HTMLTableHeaderCellElement>, params: { direction: SortDirection }) => void;
};

export type FProps = Props & Omit<React.ComponentProps<"th">, "onClick">;

const Th: React.FC<FProps> = ({ sort, onClick, children, ...props }) => {
  const [sortDirection, setSortDirection] = useState<SortDirection | "">("");
  const [classNames, rest] = useClassnames("table__thead__th", props);

  const handleClick = (e: React.MouseEvent<HTMLTableCellElement>) => {
    let newDirection: SortDirection = sortDirection || "asc";

    if (newDirection === "asc") newDirection = "desc";
    else newDirection = "asc";

    setSortDirection(newDirection);
    if (onClick) onClick(e, { direction: newDirection });
  };

  return (
    <th
      className={classNames}
      aria-sort={!sort ? "none" : sortDirection === "desc" || sortDirection === "" ? "descending" : "ascending"}
      onClick={e => handleClick(e)}
      data-testid="table-th"
      {...rest}
    >
      <div>
        {children}
        {sort && (sortDirection === "" ? <ChevronUpAndDown /> : sortDirection === "asc" ? <ChevronUp /> : <ChevronDown />)}
      </div>
    </th>
  );
};

Th.displayName = "TableTh";

export default Th;
