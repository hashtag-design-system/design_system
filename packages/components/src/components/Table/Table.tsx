import React, { useEffect, useRef, useState } from "react";
import { TableContextProvider } from "../../utils/contexts";
import { useClassnames } from "../../utils/hooks";
import TBody from "./TBody";
import Td from "./Td";
import Th from "./Th";
import THead from "./THead";
import Tr from "./Tr";

export const TableSelectionInputs = ["checkbox", "radio"] as const;

// TODO: Add the `Shift + click` <Checkbox /> keydown handler
export type TableSelectionInputsTableType = { id: string; isChecked: boolean; header: boolean };
export type TableSelectionInputType = typeof TableSelectionInputs[number];

export type Props = {
  extraColumn?: {
    component: TableSelectionInputType;
    withBorderRight?: boolean;
    selectedRows?: (row: TableSelectionInputsTableType[]) => void;
  };
};

export type FProps = Props & React.ComponentProps<"table">;

type SubComponents = {
  Th: typeof Th;
  Tr: typeof Tr;
  Td: typeof Td;
  THead: typeof THead;
  TBody: typeof TBody;
};

const Table: React.FC<FProps> & SubComponents = ({ extraColumn, children, ...props }) => {
  const [selectionInputs, setSelectionInputs] = useState<TableSelectionInputsTableType[]>([]);
  const [classNames, rest] = useClassnames("table", props);
  const ref = useRef<HTMLTableElement>(null);

  const handleClick = (e: React.ChangeEvent<HTMLInputElement> | React.MouseEvent<HTMLInputElement>, header = false) => {
    const { currentTarget } = e;
    const id = currentTarget.id;
    // Opposite because state has not been updated
    const checked = currentTarget.value === "false" ? true : false;

    if (extraColumn?.component === "checkbox") {
      setSelectionInputs(prevData =>
        prevData.map(input => {
          if (header) {
            return {
              ...input,
              isChecked: checked,
            };
          } else if (input.id === id) {
            return {
              ...input,
              isChecked: checked,
            };
          } else {
            return input;
          }
        })
      );
    } else if (extraColumn?.component === "radio") {
      setSelectionInputs(prevData =>
        prevData.map(input => {
          // checked !== true is essential, otherwise when checked === false,
          // it will set checked === true for all the other <RadioButton /> components
          if (input.id === id || checked !== true) {
            return {
              ...input,
              isChecked: checked,
            };
          } else {
            return {
              ...input,
              isChecked: !checked,
            };
          }
        })
      );
    }
  };

  useEffect(() => {
    if (extraColumn && selectionInputs.length > 0) {
      const { selectedRows } = extraColumn;
      if (selectedRows) {
        selectedRows(selectionInputs);
      }
    }
  }, [extraColumn, selectionInputs]);

  return (
    <TableContextProvider value={{ extraColumn, selectionInputs, setSelectionInputs, handleClick }}>
      <table ref={ref} className={classNames} data-testid="table" {...rest}>
        {children}
      </table>
    </TableContextProvider>
  );
};

Table.Tr = Tr;
Table.Th = Th;
Table.Td = Td;
Table.THead = THead;
Table.TBody = TBody;

export default Table;
