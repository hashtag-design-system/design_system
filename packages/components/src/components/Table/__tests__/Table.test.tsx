import { render, screen } from "@testing-library/react";
import { SortDirection } from "../../../typings";
import { useSortableData } from "../../../utils/hooks";
import Table, { TableFProps, TableThProps, TableTrFProps } from "../index";

export type TableTestInitialDataType = {
  id: number;
  amount: number;
  test: boolean;
  username: string;
};
export const tableTestInitialData: TableTestInitialDataType[] = [
  { id: 1, amount: 0.75, test: true, username: "#hashtag" },
  { id: 2, amount: 1.0, test: false, username: "me" },
  { id: 3, amount: 0.65, test: false, username: "skg" },
  { id: 4, amount: 1.5, test: true, username: "spoon" },
  { id: 5, amount: 1.25, test: false, username: "georgekrax" },
];

export type TableTestWrapperFProps = Omit<TableFProps, "onClick"> & TableThProps & { header?: boolean } & Pick<TableTrFProps, "state">;

export const TableTestWrapper: React.FC<TableTestWrapperFProps> = ({
  state = "default",
  header = true,
  sort = true,
  extraColumn,
  onClick,
  ...props
}) => {
  const { data, setSort } = useSortableData<TableTestInitialDataType>(tableTestInitialData);

  const handleClick = (
    e: React.MouseEvent<HTMLTableHeaderCellElement>,
    direction: SortDirection,
    key: keyof TableTestInitialDataType
  ) => {
    if (sort) {
      setSort({ direction, key });
    }
    if (onClick) onClick(e, { direction });
  };

  return (
    <Table
      extraColumn={
        extraColumn ? { ...extraColumn, selectedRows: row => extraColumn.selectedRows && extraColumn.selectedRows(row) } : undefined
      }
      {...props}
    >
      {header && (
        <Table.THead>
          <Table.Tr>
            <Table.Th sort={sort} onClick={(e, { direction }) => handleClick(e, direction, "id")}>
              ID
            </Table.Th>
            <Table.Th>Amount</Table.Th>
            <Table.Th sort={sort} onClick={(e, { direction }) => handleClick(e, direction, "test")}>
              Test
            </Table.Th>
            <Table.Th sort={sort} onClick={(e, { direction }) => handleClick(e, direction, "username")}>
              Username
            </Table.Th>
          </Table.Tr>
        </Table.THead>
      )}
      <Table.TBody>
        {data.map(({ id, amount, test, username }, i) => {
          return (
            <Table.Tr state={i === 1 ? state : undefined} key={i}>
              <Table.Td>{id}</Table.Td>
              <Table.Td>{amount}</Table.Td>
              <Table.Td>{String(test)}</Table.Td>
              <Table.Td>{username}</Table.Td>
            </Table.Tr>
          );
        })}
      </Table.TBody>
    </Table>
  );
};

describe("<Table />", () => {
  test("default behaviour", () => {
    render(<Table />);
    const table = screen.getByTestId("table");

    expect(table).toBeVisible();
    expect(table).toHaveAttribute("class");
    expect(table.tagName.toLowerCase()).toBe("table");
    expect(table.children).toHaveLength(0);
  });
  test("with children", () => {
    render(<TableTestWrapper />);
    const table = screen.getByTestId("table");

    expect(table).toMatchSnapshot();

    const children = table.children;
    const thead = screen.getByTestId("table-thead");
    const tbody = screen.getByTestId("table-tbody");

    expect(children).toHaveLength(2);
    expect(children).toContain(thead);
    expect(children).toContain(tbody);

    expect(thead.children).toHaveLength(1);
    expect(tbody.children).toHaveLength(5);
  });
});
