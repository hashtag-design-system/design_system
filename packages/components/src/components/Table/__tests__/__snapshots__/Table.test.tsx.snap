// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> with children 1`] = `
<table
  class="table"
  data-testid="table"
>
  <thead
    class="table__thead"
    data-testid="table-thead"
  >
    <tr
      class="table__tr "
      data-testid="table-tr"
    >
      <th
        aria-sort="descending"
        class="table__thead__th"
        data-testid="table-th"
      >
        <div>
          ID
          <svg
            class="icon chevron_up_and_down"
            data-testid="icon"
            fill="none"
            height="16"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 9l-5-6-5 6M17 15l-5 6-5-6"
              stroke="#000"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </th>
      <th
        aria-sort="none"
        class="table__thead__th"
        data-testid="table-th"
      >
        <div>
          Amount
        </div>
      </th>
      <th
        aria-sort="descending"
        class="table__thead__th"
        data-testid="table-th"
      >
        <div>
          Test
          <svg
            class="icon chevron_up_and_down"
            data-testid="icon"
            fill="none"
            height="16"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 9l-5-6-5 6M17 15l-5 6-5-6"
              stroke="#000"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </th>
      <th
        aria-sort="descending"
        class="table__thead__th"
        data-testid="table-th"
      >
        <div>
          Username
          <svg
            class="icon chevron_up_and_down"
            data-testid="icon"
            fill="none"
            height="16"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 9l-5-6-5 6M17 15l-5 6-5-6"
              stroke="#000"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </th>
    </tr>
  </thead>
  <tbody
    class="table__tbody"
    data-testid="table-tbody"
  >
    <tr
      class="table__tr "
      data-testid="table-tr"
    >
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        1
      </td>
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        0.75
      </td>
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        true
      </td>
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        #hashtag
      </td>
    </tr>
    <tr
      class="table__tr "
      data-testid="table-tr"
    >
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        2
      </td>
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        1
      </td>
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        false
      </td>
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        me
      </td>
    </tr>
    <tr
      class="table__tr "
      data-testid="table-tr"
    >
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        3
      </td>
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        0.65
      </td>
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        false
      </td>
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        skg
      </td>
    </tr>
    <tr
      class="table__tr "
      data-testid="table-tr"
    >
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        4
      </td>
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        1.5
      </td>
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        true
      </td>
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        spoon
      </td>
    </tr>
    <tr
      class="table__tr "
      data-testid="table-tr"
    >
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        5
      </td>
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        1.25
      </td>
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        false
      </td>
      <td
        class="table__tbody__td"
        data-testid="table-td"
      >
        georgekrax
      </td>
    </tr>
  </tbody>
</table>
`;
